File: 240626/단순한 동전 챙기기/collect-coins-easy.cpp
#include <iostream>
#include <algorithm>
#include <tuple>
#include <cmath>
#include <climits>
#include <vector>
using namespace std;

#define MAX_N 20
#define COIN_NUM 9

int n, m=3;
char grid[MAX_N][MAX_N];

vector<pair<int, int>> coin_pos;
vector<pair<int, int>> selected_pos;

pair<int, int> start_pos;
pair<int, int> end_pos;

int ans = INT_MAX;

int Dist(pair<int, int> a, pair<int, int> b) {
    int ax, ay;
    tie(ax, ay) = a;

    int bx, by;
    tie(bx, by) = b;

    return abs(ax-bx)+abs(ay-by);
}

int Calc() {
    int num_moves = Dist(start_pos, selected_pos[0]);

    for(int i=0; i<m-1; i++) {
        num_moves += Dist(selected_pos[i], selected_pos[i+1]);
    }

    num_moves += Dist(selected_pos[m-1], end_pos);

    return num_moves;
}

void FindMinMoves(int cur_idx, int cnt) {
    if(cnt == m) {
        ans = min(ans, Calc());
        return;
    }

    if(cur_idx == (int)coin_pos.size()) return;

    FindMinMoves(cur_idx+1, cnt);

    selected_pos.push_back(coin_pos[cur_idx]);
    FindMinMoves(cur_idx+1, cnt+1);
    selected_pos.pop_back();
}


int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n;

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> grid[i][j];
            if(grid[i][j] == 'S') start_pos = make_pair(i, j);
            if(grid[i][j] == 'E') end_pos = make_pair(i, j);
        }
    }

    for(int num=1; num<=COIN_NUM; num++) {
        for(int i=0; i<n; i++) {
            for(int j=0; j<n; j++) {
                if(grid[i][j] == num+'0') {
                    coin_pos.push_back(make_pair(i, j));
                }
            }
        }
    }

    FindMinMoves(0, 0);

    if(ans == INT_MAX) ans = -1;

    cout << ans;
    return 0;
}

File: 250703/단순한 동전 챙기기/collect-coins-easy.py
import sys

COIN_NUM = 9
INT_MAX = sys.maxsize

n = int(input())
m = 3

grid = [
    input()
    for _ in range(n)
]

coin_pos = [
    (i, j)
    for num in range(1, COIN_NUM+1)
    for i in range(n)
    for j in range(n)
    if grid[i][j] == str(num)
]
start_pos = [
    (i, j)
    for i in range(n)
    for j in range(n)
    if grid[i][j] == 'S'
][0]

end_pos = [
    (i, j)
    for i in range(n)
    for j in range(n)
    if grid[i][j] == 'E'
][0]

selc_pos = []

ans = INT_MAX

def dist(a, b):
    (x1, y1), (x2, y2) = a, b
    return abs(x2-x1) + abs(y2-y1)

def calc():
    num_moves = dist(start_pos, selc_pos[0])
    for i in range(m-1):
        num_moves += dist(selc_pos[i], selc_pos[i+1])
    num_moves += dist(selc_pos[m-1], end_pos)

    return num_moves

def find_min_move(idx, cnt):
    global ans

    if cnt == m:
        ans = min(ans, calc())
        return
    
    if idx == len(coin_pos):
        return
    
    find_min_move(idx+1, cnt)

    selc_pos.append(coin_pos[idx])
    find_min_move(idx+1, cnt+1)
    selc_pos.pop()


find_min_move(0, 0)
print(ans if ans != INT_MAX else -1)